src/states.js

```diff
@@ -0,0 +1,370 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { bytesAreEqual, BufferWriter } from './utils.js';
+import { TLSError, ALERT_DESCRIPTION } from './alerts.js';
+import { getRandomBytes, HASH_LENGTH } from './crypto.js';
+import {
+  ClientHello,
+  ServerHello,
+  EncryptedExtensions,
+  Finished
+} from './messages.js';
+
+// The length of the data for PSK binders at the end of the ClientHello.
+// We only support a single PSK, so it's the length of the hash plus one
+// for rendering it as a variable-length byte array, plus two for rendering
+// the variable-length list of PSK binders.
+const PSK_BINDERS_SIZE = HASH_LENGTH + 1 + 2;
+
+//
+// State-machine for TLS Handshake Management.
+//
+// Internally, we manage the TLS connection by explicitly modelling the
+// client and server state-machines from RFC8446.  You can think of
+// these `State` objects as little plugins for the `Connection` class
+// that provide different behaviours of `send` and `receive` depending
+// on the state of the connection.
+//
+
+class State {
+
+  constructor(conn) {
+    this.conn = conn;
+  }
+
+  async initialize() {
+    // By default, nothing to do when entering the state.
+  }
+
+  async sendApplicationData(bytes) {
+    // By default, assume we're not ready to send yet
+    // and just let the data queue up for a future state.
+    this.conn._sendBuffer.push(bytes);
+  }
+
+  async recvApplicationData(bytes) {
+    throw new TLSError(ALERT_DESCRIPTION.UNEXPECTED_MESSAGE);
+  }
+
+  async recvHandshakeMessage(msg) {
+    throw new TLSError(ALERT_DESCRIPTION.UNEXPECTED_MESSAGE);
+  }
+
+  async recvAlertMessage(err) {
+    switch (err.description) {
+      case ALERT_DESCRIPTION.USER_CANCELLED:
+        // It's not obvious from the spec whether we should do anything here,
+        // or just wait for the following `CLOSE_NOTIFY` alert.
+        break;
+      case ALERT_DESCRIPTION.CLOSE_NOTIFY:
+        this.conn._closeForRecv();
+        break;
+      default:
+        return await this.handleError(err);
```

Comment:
Note to self: this won't actually throw the error, but it should!

Reply:
Maybe you should name the function `error` when you have it throw.

Reply:
yep; in my head it's something like `handleAndRethrowError` to make it super clear what will happen.